Improve the AsciiDoc reporter

johthor opened the pull request:
Improve the AsciiDoc reporter

Some time ago in one of our projects we introduced documentation via AsciiDoc and later also some JGiven based behavior tests. Unfortunately the interactive HTML report was not compatible with the AsciiDoc based documentation.

When we found the AsciiDoc reporter via #54 we had high hopes to be able to include the produced AsciiDoc report into our documentation. Only to realize that data tables as step parameters where not supported while most of our scenarios
relied heavily on those data tables.

So some month ago we started with improving the AsciiDoc reporter by adding support for all the features we needed. Now we have reached a development stage where most of our needs are met but we don't know how to continue and if the AsciiDoc reporter is even still a wanted feature.

An example for the generated AsciiDoc files is pushed to https://github.com/eCubeGmbH/JGiven/tree/asciiDocReport under `current` while `goal` contains some ideas what could be possible.

As discussed in #54 we removed the `ReportModelHandler` again and reverted to a `ReportModelVisitor` based approach. Where the `AsciiDocReportBlockConverter` is reponsible for transforming data from the model into AsciiDoc text blocks that are later only printed into the corresponding files.

The following features and properties are supported for now:

- overall report
  - [x] statistics summary table
- features
  - [x] class name / name (optional)
  - [x] description (optional)
  - [x] mini statistic
- scenarios
  - [x] description
  - [x] extended description (optional)
  - [x] status
  - [x] duration
  - [x] tags (optional)
- standalone scenario cases
  - [x] case number
  - [x] description (optional)
  - [x] status icon
  - [x] duration
  - [x] error message (optional)
  - [x] stack trace (optional)
- scenario cases as table
  - [x] case number
  - [x] parameters
  - [x] description (optional)
  - [x] status icon
  - [x] duration
  - [x] error message (optional)
  - [x] stack trace (optional)
- steps
  - [x] status
  - [x] duration
  - [x] description
  - [x] section titles
  - [x] nested steps
  - [ ] attachments _not to be added in this PR, it is already quite big_

l-1squared commented:
Great! :)
Many thanks in advance for that contribution! I will have a look at it as soon as I find the time. (Probably next week)

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/CasesTableImpl.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Override
+    public List<Row> rows() {
+        List<Row> rows = Lists.newArrayList();
```
Could the Stream API be used here?

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/CasesTableImpl.java:
personal note: The contents fo this file come from the `AbstractReportModelHandler`

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/CasesTable.java:
I am a bit uncertain about this and the `ReportBlockConverter` interface. Right now they are only used for the AsciiDoc generator and I have not yet figured out if they could be used for the other report outlets.

johthor replied:
Regarding the `CasesTable` as you have already discovered it was part of the remove `AbstractReportModelHandler`. We found it to provide a simple interface for collecting multiple cases into a table. This greatly simplifies reporting on those cases and should be independent from the reporting format used. 

But until now we did not really improve on the implementation of the `CasesTableImpl`. It might be a better approach to collect the table information via a visitor like it is done in the `StatisticsCalculator`. The collection of data should be independent from any reporting format.

The `ReportBlockConverter` on the other hand is part of our reinterpretation of the `AbstractReportModelHandler` class and the `ReportModelHandler` which seemed to provide a format agnostic way to report.  Our idea was for the `ReportBlockConverter` to provide a clear distinction between traversing the model and transforming the information into a specific format. We imagined this to be a starting point for all markup languages where content can be understood as a set of text blocks divided by a block divider like new lines or empty lines. The code might not have communicated the idea in its fullest but a Markdown or reStructured Text reporters could follow the same pattern.

l-1squared reviewed (commented):
Unfortunately I did not manage to finish reviewing your PR. There is a lot to look at and a lot to think about here. I have left some initial thoughts though.

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportBlockConverterTest.java:
```diff
@@ -0,0 +1,647 @@
+            "",
+            "icon:check-square[role=green] (3s 0ms)",
+            "",
+            "+++Best scenario ever!!!+++",
```
what kind of formatting do these plusses do?

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportBlockConverterTest.java:
```diff
@@ -0,0 +1,647 @@
+    }
+
+    @Test
+    public void convert_case_header_with_one_parameter() {
```
There doesn't seem to be a diffference between this and the following test.

l-1squared replied:
Also I fail to find the parameters in this test

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportBlockConverterTest.java:
```diff
@@ -0,0 +1,647 @@
+        List<Word> words = ImmutableList.of(Word.introWord("given"), new Word("a coffee machine"));
+
+        // act
+        String block = converter.convertStepBlock(0, words, StepStatus.PASSED, 3899, null, false);
```
I would put the number `3899` in a global constant name `DURATION` (or DURATION_IN_NANOS). The number doesn't actually matter, but a reader would have an easier time determining its significance

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportBlockConverter.java:
```diff
@@ -0,0 +1,432 @@
+
+class AsciiDocReportBlockConverter implements ReportBlockConverter {
+
+    private static final String NEW_LINE = System.getProperty("line.separator");
```
there's `System.lineSeparator()` which would be even more direct.

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportBlockConverter.java:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    private static String escapeArgumentValue(final String value) {
+        // TODO Is this really necessary?
```
that would be great to find out :)

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportBlockConverter.java:
```diff
@@ -0,0 +1,432 @@
+    private static String toAsciiDocTag(final ExecutionStatus executionStatus) {
+        switch (executionStatus) {
+            case SCENARIO_PENDING:
+                // fall through
```
since that is a standard Java feature, I would omit this comment.

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportBlockConverter.java:
```diff
@@ -0,0 +1,432 @@
+            : generate(() -> "1").limit(columnCount).collect(joining(","));
+    }
+
+    private static String toAsciiDocTag(final ExecutionStatus executionStatus) {
```
How about putting these enum conversion methods in a package-private `EnumValuesMappingUtil` so that this class does not need to concern itsself with the mapping details?

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportBlockConverter.java:
```diff
@@ -0,0 +1,432 @@
+        }
+    }
+
+    private static Optional<String> toHumanReadableDuration(final long nanos) {
```
I take issue with this method, because it doesn't exactly announce what it does. (and one may also argue that it does two things), because the method name, in conjunction with the return type suggests that the method will attempt a conversion, which may fail, when in fact it makes a decision whether it should convert AND converts the input.

Unfortunately, I don't have a nifty alternative at hand right now :(

l-1squared commented on jgiven-tests/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGeneratorTest.java:
```diff
@@ -1,50 +1,54 @@
-                "...." + System.lineSeparator() +
-                        content + System.lineSeparator() +
-                        "....");
+        then().the_asciidoc_reporte_$_exists("features/Test.asciidoc")
```
I kindly want to ask you to fix this old spelling mistake

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGenerator.java:
```diff
@@ -1,196 +1,186 @@
-            writer.println( description );
-            writer.println();
-        }
+        try (PrintWriter printWriter = PrintWriterUtil.getPrintWriter(new File(targetDir, "allScenarios.asciidoc"))) {
```
could these `try` statements be split into several methods, like `writeIndex`, `writeGlobalStatistics` etc.?

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGenerator.java:
```diff
@@ -1,196 +1,186 @@
-            writeReportModelToFile( reportModelFile.model, reportModelFile.file );
-        }
-        generateIndexFile();
+    public AsciiDocReportGenerator() {
```
what is the significance of putting this into the constructor, vs just assigning it at field declaration?

l-1squared reviewed (commented):
This is a complex PR :) 
But at least I think I managed to have glance at all the files by now.

l-1squared reviewed (commented):
This is a complex PR :) 
But at least I think I managed to have glance at all the files by now.

johthor reviewed (commented):
@l-1squared I went through all your comments and incorporated all your suggestions.

Some of them I couldn't believe I missed when creating the PR :grin:

johthor commented:
I've now reworked the `CasesTable` interface into a simple class without a `CasesTableImpl` and moved the calculation into `CasesTableCalculator`. I'm not to sure about the package `report.model` but every other sub-package seemed equally wrong.

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGenerator.java:
```diff
@@ -1,196 +1,213 @@
-        AsciiDocReportModelVisitor( PrintWriter printWriter ) {
-            this.writer = printWriter;
+    private boolean prepareDirectories() {
+        targetDir = config.getTargetDir();
```
do we have a test that`targetDir` won't be `null` before it is used in other methods?

johthor replied:
No there was no test until now and I'm not sure there is even now. I've added some tests but to be honest the different usage patterns with `generate()`, `setConfig` and `loadReportModel` being public and sometimes called from other classes did confuse me a little bit.

I would have preferred for those to not be public.

I hope that `targetDir` being `null` is now handled properly.

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGenerator.java:
```diff
@@ -1,196 +1,213 @@
-        @Override
-        public void stepEnd() {
-            writer.println( "+" );
+    private void writeFailedScenariosFiled() {
```
I presume the last word is `Files`

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportModelVisitorTest.java:
```diff
@@ -0,0 +1,162 @@
+        reportModel.accept(reportModelVisitor);
+
+        // then
+        Assertions.assertThat(reportModelVisitor.getResult())
```
could be a static import

l-1squared reviewed (commented):
Only  managed to have a short glance today (mostly verified that we can run the tests again). But it looks mostly fine. Maybe we're starting to converge now.

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/model/CasesTable.java:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Represents one case of a scenario.
+     */
+    public static class CaseRow {
```
yet another case, where a newer java version would have been o so handy.

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/model/CasesTable.java:
```diff
@@ -0,0 +1,139 @@
+        /**
+         * The duration of the case.
+         */
+        public long getDurationInNanos() {
```
since all other getters are not prefixed, maybe this one should not be either.

l-1squared commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/model/CasesTableCalculator.java:
I like this class, it makes the report model visitor look so neat :)

l-1squared reviewed (commented):
I managed to get through the productive code today and I like it. 

If I am not mistaken, however, then we need to have a go at the gradle plugin, to make the AsciiDoc report available to gradle users as well

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportConfigTest.java:
```diff
@@ -0,0 +1,37 @@
+
+import org.junit.Test;
+
+public class AsciiDocReportConfigTest {
```
How about another test with an input "--sourceDir=". We currently fail gracefully and print the help message, which I think is good behavior.

johthor replied:
The `ConfigOptionParser` calls `System.exit()` in this case and currently I'm not sure how to test for that case without introducing a lot of additional complexity.

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGeneratorTest.java:
```diff
@@ -0,0 +1,76 @@
+package com.tngtech.jgiven.report.asciidoc;
```
Shouldn't we at least have some test here that checks that we have actually generated a report?

johthor replied:
There is the identical named `jgiven-tests/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGeneratorTest.java` which checks that the five meta files and one feature file is generated. Which was enough for me. Or to be honest I didn't want to handle files and file contents in the tests.

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportModelVisitorTest.java:
```diff
@@ -0,0 +1,163 @@
+                        "convertScenarioFooterBlock"));
+    }
+
+    private static class MyFakeReportBlockConverter implements ReportBlockConverter {
```
I am curious why you chose this implementation over just using mockito. Is there a specific reason?

johthor replied:
Mockito was not part of the test dependencies of the `jgiven-core` project and I didn't want to add it if it wasn't already.
I presumed there was a reason for it to not be added. So I implemented my own _order of invocation verifier_.

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportModelVisitorTest.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Test
+    public void visitTheReportModelWithTwoSections() {
```
shouldn't this be spelt `visitAScenarioWithTwoScenarioCases`? Speaking of which, checking that two Scenarios in a report are handled gracefully might also be a good idea

johthor replied:
Yeah the test did check multiple things. Now there are more tests that hopefully only check for exactly one situation.

l-1squared reviewed (commented):
More review. I'm halfway through the test files now :)

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocStepBlockConverterTest.java:
```diff
@@ -0,0 +1,218 @@
+    @Test
+    @DataProvider({"PASSED, check-square[role=green]", "FAILED, exclamation-circle[role=red]",
+        "SKIPPED, step-forward[role=silver]", "PENDING, ban[role=silver]"})
+    public void convert_step_from_unsuccessful_scenario_case(final StepStatus stepStatus, final String icon) {
```
Given that the data provider also includes "passed", "unsuccesful" is inapproproate here. How about "converts_scenario_cases_step_status"?

johthor replied:
No, the unsuccessful does not described the status of the steps execution but the execution status of the scenario case it is included in.
In successful scenarios and scenario cases there is no need to print every step's execution status as they should all be `passed` only for non successful scenario cases the steps status should be reported.

johthor replied:
But that comment showed me that the status reporting was not implemented correctly for all parameter types ;(

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/model/CasesTableCalculatorTest.java:
```diff
@@ -0,0 +1,173 @@
+        assertThat(casesTable.placeholders()).containsExactly("foo", "bar");
+        assertThat(casesTable.rows())
+                .hasSize(2)
+                .satisfiesExactly(
```
Oh! I didn't know that feature.

l-1squared reviewed (commented):
@johthor  I made it through! 
I'm happy about this PR.

l-1squared commented on jgiven-gradle-plugin/src/main/java/com/tngtech/jgiven/gradle/internal/AbstractJGivenReportImpl.java:
```diff
@@ -57,7 +57,7 @@ public AbstractReportGenerator createGenerator() {
             conf.setTitle( getTitle() );
         }
-        conf.setTargetDir( getDestination() );
+        conf.setTargetDir( getOutputLocation().getAsFile().get() );
```
another gradle deprecation?

johthor replied:
yes.

l-1squared replied:
thanks for resolving

l-1squared reviewed (commented):
Looks fine at a glance. Will do a final sweep on monday

johthor commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportGenerator.java:
```diff
@@ -1,196 +1,226 @@
-            for( ScenarioDataTable.Row row : scenarioDataTable.rows() ) {
-                writer.print( "| " + row.nr() );
+        final AsciiDocReportModelVisitor visitor = new AsciiDocReportModelVisitor(blockConverter, statistics);
+        reportModelFile.model.accept(visitor);
```
I cant't answer the question for that code from 2014. Until now I tried not to change anything I did not write myself or absolutely had to modify.

But I will gladly change that class if you want me to.

l-1squared replied:
Seems you did, and that was a good little refactoring.

johthor commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocScenarioBlockConverterTest.java:
```diff
@@ -0,0 +1,122 @@
+        List<String> tagNames = new ArrayList<>();
+
+        // when
+        String block = converter.convertScenarioHeaderBlock("my first scenario", status, 1000300000L, tagNames, null);
```
Those are 300000 nanoseconds so there are no milliseconds to be displayed. I've now introduced underscores to better communicate the different time units and extracted variables.

l-1squared commented on jgiven-core/src/test/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportModelVisitorTest.java:
```diff
@@ -0,0 +1,296 @@
+        // given
+
+        ReportModel report = mkReport(
+                mkScenario("Scenario One", false, mkScenarioCase(
```
Its really rare to see helper /shortcut methods in a test that work that well. Nice :)

l-1squared reviewed (commented):
Hi, 
so codewise I think we're fine. 
But I noticed that there are failing test (for me locally) in core and jgiven-tests. 
The one in core is somewhat my fault as I asked you to write one with bogus input for the AsciiDocConigBuilder (or a similar name). I guess we need to scrap that. 

then in test, my candidate for most annoying test in the codebase seems to fail for you, which seems to stem from the fact that you are not using the current master as base. 

Finally, I took a look and the ascii-report generated for JGiven tests and I noticed two things.
* in com.tngtech.jgiven.report.analysis.ArgumentAnalyzerTest.asciidoc the tag of the first test doesn't get processed properly and loses its number. I'd say that this should be fixed, before this awesome is released
* Secondly I think that the hints as such (i.e. text written in cursive, see same test), may be only recognizeable to the initiated. I was wandering wheter placing a "hint" or "meaning" before that hint and/or putting the hint in parentheses will help here.

johthor commented on jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocSnippetGenerator.java:
```diff
@@ -0,0 +1,62 @@
+            result.add(":leveloffset: -1");
+        }
+
+        for (final String fileName : featureFiles) {
```
Yeah sure.

johthor commented:
I must already have removed the failing `AsciiDocReportConfigTest` again after discovering that `System.exit()` was involved :grin: .

And I now have merged the current master. I hope that will take care of some of those failures.

I've introduced a :heavy_plus_sign: in front of the extended step descriptions to hopefully communicate the meaning better. I didn't like the idea of introducing text for explanation or parentheses. The icon also got a title that should display on hover. Unfortunately there is no tool tip in AsciiDoc like in the HTML report.

And I will try fix the handling of tags. I did know that the implementation is lacking there.

johthor commented:
For me locally only tests from `Html5AppTest` from `jgiven-tests` are failing and I don't fully understand why they are failing.

l-1squared commented:
> I must already have removed the failing `AsciiDocReportConfigTest` again after discovering that `System.exit()` was involved 😁 .
> 
> And I now have merged the current master. I hope that will take care of some of those failures.
> 
> I've introduced a ➕ in front of the extended step descriptions to hopefully communicate the meaning better. I didn't like the idea of introducing text for explanation or parentheses. The icon also got a title that should display on hover. Unfortunately there is no tool tip in AsciiDoc like in the HTML report.
> 
> Handling tags with values and without values was less complex than I thought. And now there is also support for `cssClass`.

As I said, the test is a total nuissance. Yesterday (bevore merging) it failed on my setup because it was still hardwired to the chrominum driver, which currently seems to have a bug that prevents chromium (and maybe all chromium based browsers) to access any websit if not some "origins allowed is everything" header is set. 
I (and I think the CI) don't see these problems, because the test first tries out which browsers are present in the order firefox -> chrome -> edge. (Reduces reproducability, but then again the choice of browser should not matter for the test proper and what information is there to be had when you fail, because the desired browser was not available). So maybe you fail because the test's browser selecter ends up using chrome whereas the CI and I end up using firefox.

l-1squared reviewed (dismissed):
I see that you have not yet completed your list of features that you'd like to have. 
But maybe it makes sense to add these things as a new PR

johthor commented:
YES, that would be the better choice.

Printing the parameters for standalone scenarios was done in the original implementation, to behave like the HTML report. But I decided to remove it as the standalone cases do print the parameters values inline anyways.

The duration for scenario cases I thought would not be shown in the HTML report which seems to be dependent on the situation. I think I want to still add it, it shouldn't be a big change.

The Attachments I won't add in this PR, as I'm not sure if it is worth to re-implement the `HTML5AttachmentGenerator` which is unfortunately only part of the `jgiven-html5-report` project.

johthor commented:
And I'm done. :grin: 

I've updated the PR description accordingly and even improved the rendering of tables a little bit.
I don't want to get any more _quick_ ideas in this PR.

l-1squared reviewed (commented):
Hi,  

I have noticed some issues while running the test on my local machine:
* There are some failing tests in jgiven-core on my machine:
** AsciiDocScenarioCaseBlockConverterTest#convert_scenario_case_footer_without_stacktrace
** AsciiDocTableBlockConverterTEst#conert_cases_table_with_errors
** AsciiDocStepBlockConverterTest

This seems to be an issue with line breaks which  for some reason I can see Unix line breaks  on my Windows machine, but I have also noticed, that at least some tests just have far too few of them.

Another, more troublesome matter is that I did not find a duration in the asciidocReport, where I expected it. Case  in point: jgiven-tests, HTML App->Newlines are detected in formatted values and shown as multiline text, Step "When the HTML Report Generator is executed" and the following one. They do have durations on them in the HTML App, but not in the ascii doc report. 
Do you see the same behavior?

johthor commented:
Hi @l-1squared,

unfortunately I completely forgot about the line ending problem with Windows and Linux/MacOS and did not use the `System.lineSeparator()` in all places. That should hopefully be fixed now. I'm missing a Windows development machine so I can't be sure. I did use `\n` in test inputs but the code checked for `System.lineSeparator()` which was `\r\n` on your machine.

Other parts of the code solved this problem by using `actual.replaceAll(System.getProperty("line.separator"), "\n")`. I'm not sure which way is simpler to understand.

Regarding the step duration I could reproduce the outcome => I did misunderstand the display of durations in the HTML report again. ;(

I've also reworked that part. Now step status and step duration may be displayed independent from one another. Step status is only displayed if the surrounding scenario (case) was unsuccessful. Step duration is only displayed if it is above 10,000,000 nanoseconds or 10 milliseconds.

The status and duration of scenarios and scenario cases is always displayed.

l-1squared reviewed (approved):
@johthor  I'll let you do the honors

(do you actually have permission to do so?)

johthor commented:
I would love to click the BIG BUTTON of happiness but I may only _close with comment_ or _comment_ on this PR.

So you will have to merge the PR as I don't have write permissions.

l-1squared commented:
> @l-1squared I would love to click the BIG BUTTON of happiness but I may only _close with comment_ or _comment_ on this PR.
> 
> So you will have to merge the PR as I don't have write permissions.

Done! (Took a while, I was on a business trip)
